### ISSUE TITLE
Add public API for Dataset._copy_listed

### ISSUE THREAD

**Issue by nbren12** [pydata/xarray#3894]
In my data pipelines, I have been repeatedly burned using indexing notation to grab a few variables from a dataset in the following way:
```
ds = xr.Dataset(...)
vars = ('a' , 'b', 'c')
ds[vars] # this errors
ds[list(vars)] # this is ok
```
Moreover, because `Dataset__getitem__` is type unstable,  it makes it hard to detect this kind of error using mypy, so it often appears 30 minutes into a long data pipeline. It would be great to have a type-stable method that can take any sequence or variable names and return the Dataset consisting of those variables and their coordinates only. In fact, this method already [exists](https://github.com/pydata/xarray/blob/6378a711d50ba7f1ba9b2a451d4d1f5e1fb37353/xarray/core/dataset.py#L1123), but it currently not public API. Could we make it so? Thanks.

**Comment by max-sixty** [comment]
Would that be different from ensuring the input is a list?

> Moreover, because `Dataset__getitem__` is type unstable,

I very much empathize with the pain from methods being type unstable; indeed I think that's one of the biggest benefits of xarray over pandas. Here, it's stable _over the same typed inputs_. i.e. if supplied with a list, it returns with a dataset, otherwise it returns a DataArray. (or am I missing something?)

> it makes it hard to detect this kind of error using mypy

Is there a way in mypy we could use something like `overload` to specify the above contract here, as an alternative to another method?

**Comment by dcherian** [comment]
What's the reasoning for not returning a Dataset when `__getitem__` is passed an Iterable like `_copy_listed`?

**Comment by shoyer** [comment]
I agree, this API is too overloaded. It would be better to have an explicit method for subsetting `Dataset` variables. Maybe `subset` or `sel_vars`?

In early versions of xarray (back when it was called xray), we actually had a `select` method but I was concerned it was too confusing with indexing: http://xarray.pydata.org/en/v0.1.1/generated/xray.Dataset.select.html#xray.Dataset.select

> What's the reasoning for not returning a Dataset when `__getitem__` is passed an Iterable like `_copy_listed`?

The current check uses hashability to determine whether to try to make a DataArray. In theory, you could put a variable with the name `('a', 'b', 'c')` into a Dataset.

**Comment by nbren12** [comment]
> Is there a way in mypy we could use something like overload to specify the above contract here, as an alternative to another method?

That is correct. The output type is predictable from the inputs types. With #4144, `mypy` may have a chance at detecting this kind of error. Still, I don't know how many users use type-checking. I expect most will only discover this problem at runtime.

>  It would be better to have an explicit method for subsetting Dataset variables.

I agree. `sel_vars` is more clear IMO since `subset` could apply to the coordinates too e.g. a spatial subset.

**Comment by nbren12** [comment]
Or maybe "get" since it's a synonym of "select" that isn't overloaded with spatial indexing in the code base.

**Comment by nbren12** [comment]
NVM, `get` is already a method. Maybe we could overload it?

**Comment by shoyer** [comment]
> > It would be better to have an explicit method for subsetting Dataset variables.
> 
> I agree. `sel_vars` is more clear IMO since `subset` could apply to the coordinates too e.g. a spatial subset.

We did a similar splitting of functionality recently with `drop`, into `drop_vars` and `drop_sel`.

So this would leave us with:

- `sel`/`drop_sel` for indices
- `sel_vars`/`drop_vars` for variables

The naming doesn't have an obvious pattern here, which seems non-ideal. I can't think of anything much better at the moment, but perhaps it would help to avoid reusing `sel`. Maybe `get_vars` or `subset_vars`?

**Comment by dcherian** [comment]
I think avoiding `sel` is a good idea (but no strong thoughts here).

how about  `pick_vars` or `take_vars`?

**Comment by max-sixty** [comment]
I do think having a `Dataset` behave similarly to a `dict` / `Mapping` is generally good, and allows new users to bring  existing understandings around those data structures to the xarray data model. 

I recognize that a hashable iterable (e.g. `('a', 'b', 'c')` is an annoying corner case, though.

**Comment by max-sixty** [comment]
A level down, re the name — I thought `.vars` might be decent — but potentially it's too similar to `.variables` — which is a mapping to the actual variables (and can't take multiple selections)

**Comment by jthielen** [comment]
Would this proposal mean that subsetting variables with `__getitem__` would no longer work? If so, I'd make the humble request as a downstream user/library contributor for it to have a very generous deprecation period, since it is core functionality I rely on lot (and include in many tutorials).

**Comment by keewis** [comment]
IIUC, what we're discussing here is *adding* a new method that treats all sequences the same (`__getitem__` won't be touched, except maybe the type hints).

**Comment by shoyer** [comment]
At most, I would require using the new method if you want your code to type-check properly.

**Comment by nbren12** [comment]
I think the issue is still valid, we just couldn't think of what to name the new API.
